### PR TITLE
fix!: Reliable block header

### PR DESCRIPTION
**Pull request by brewmaster012**
# Description

This PR improves reliability of posting block headers to zetacore so that the headers form a chain connected
by the ParentHash. 

1. A BlockHeaderState variable is added to state machine to track latest block headers on each chain
2. zetaclient will use the BlockHeaderState to fetch and submit the "next" block header
3. zetacore state machine will reject block header submissions that do not have a parent already in (this adds a bit more validation to the block header chains and make it harder to inadvertently or fabricate the block headers). 

## Upgrade compatibility
Consensus breaking; no need for state migration. 

Added a smoketest to test continuity of block headers forming a chain. 

related: #1255 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [x] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/observer/keeper/msg_server_add_block_header.go
Comment (ws4charlie) [inline x/observer/keeper/msg_server_add_block_header.go:52]: Just in case, maybe we can make it more strict
```
header, found := k.GetBlockHeader(ctx, phash)
if !found {
	return nil, cosmoserrors.Wrap(types.ErrNoParentHash, "parent block header not found")
}
if header.Height+1 != msg.Height {
	return nil, cosmoserrors.Wrapf(types.ErrParentHeightMismatch, "invalid height, expected: %d, got: %d", header.Height+1, msg.Height)
}
if bytes.Compare(header.Hash, phash) != 0 {
	return nil, cosmoserrors.Wrapf(types.ErrParentHashMismatch, "parent hash mismatch, expected: %s, got: %s", msg.BlockHash, phash)
}
```

File: zetaclient/bitcoin_client.go
Comment (ws4charlie) [inline zetaclient/bitcoin_client.go:293]: Does `bn > tip` mean that zetacore is waiting for block `bn` and zetaclient is lagging behind and doesn't need to do anything in here? I meant should we return an error or just return a `nil` to skip?

**Review by ws4charlie — COMMENTED** [review]
Looks cool. Just a few comments

File: zetaclient/bitcoin_client.go
Comment (brewmaster012) [reply zetaclient/bitcoin_client.go:293]: the "tip" is supposed to be a few blocks behind what RPC returns the latest block number, essentially the "confirmed" block number. If bn > tip then we need to wait for block confirmation.

File: x/observer/keeper/msg_server_add_block_header.go
Comment (lumtis) [inline x/observer/keeper/msg_server_add_block_header.go:39]:
```suggestion
		return nil, cosmoserrors.Wrap(types.ErrBlockAlreadyExist, msg.BlockHash)
```
I think we should use error wrapping to make the error more explicit with the context of the call

File: x/observer/keeper/msg_server_add_block_header.go
Comment (lumtis) [inline x/observer/keeper/msg_server_add_block_header.go:38]: What if we have a bug and a wrong header is added?
Shouldn't we consider the capability to update an existing header?

File: x/observer/keeper/msg_server_add_block_header.go
Comment (lumtis) [inline x/observer/keeper/msg_server_add_block_header.go:43]: We should add comments to explain this condition

File: x/observer/keeper/msg_server_add_block_header.go
Comment (lumtis) [inline x/observer/keeper/msg_server_add_block_header.go:88]: `found` is defined above in the method, I think we should provide a more explicit name to make the code easier to understand

File: x/observer/keeper/msg_server_add_block_header.go
Comment (brewmaster012) [reply x/observer/keeper/msg_server_add_block_header.go:48]: done [d8eacc3](https://github.com/zeta-chain/node/pull/1374/commits/d8eacc36e3f6c2e6ec2242a8cf4a486f9f38c489)

File: common/test_data/eth_header_18495266.json
Comment (brewmaster012) [reply common/test_data/eth_header_18495266.json:1]: don't know a better place there to move to
since this is test case for common package this should probably be the appropriate place.

File: x/observer/keeper/block_header.go
Comment (brewmaster012) [reply x/observer/keeper/block_header.go]: done https://github.com/zeta-chain/node/pull/1374/commits/a383900a170fcd6f95f2541b87cca6ddaa625f1e

File: x/observer/keeper/msg_server_add_block_header.go
Comment (brewmaster012) [reply x/observer/keeper/msg_server_add_block_header.go:92]: done at the expense of slightly more read ops for the sake of readability. 
https://github.com/zeta-chain/node/pull/1374/commits/be1b98d466cda26bdadcc982fc7987f7d05ff642

File: x/observer/keeper/msg_server_add_block_header.go
Comment (ws4charlie) [reply x/observer/keeper/msg_server_add_block_header.go:38]: I was also thinking about the capability of replacing a existing block header (in store), but seems it happens only when protocol has a bug or when majority zetaclients (or EVM node providers) get corrupted, which should be sufficiently rare.

**Review by ws4charlie — APPROVED** [review]
looks good

File: x/observer/keeper/msg_server_add_block_header.go
Comment (brewmaster012) [reply x/observer/keeper/msg_server_add_block_header.go]: revised a bit

File: .github/workflows/build.yml
Comment (lumtis) [inline .github/workflows/build.yml:82]: I think we might really need to consider making the smoke test more optimized or separating the smoke test into different types

**Review by lumtis — APPROVED** [review]

**Comment by lumtis** [comment]
@brewmaster012 it will require a review from @zeta-chain/devops since `.github` has been modified

**Pull request merged**
